clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -331,13 +331,20 @@ public KafkaProducer(Properties properties, Serializer<K> keySerializer, Seriali
                 valueSerializer));
         try {
             Map<String, Object> userProvidedConfigs = config.originals();
-            this.producerConfig = config;
             this.time = time;
 
             String transactionalId = userProvidedConfigs.containsKey(ProducerConfig.TRANSACTIONAL_ID_CONFIG) ?
                     (String) userProvidedConfigs.get(ProducerConfig.TRANSACTIONAL_ID_CONFIG) : null;
 
-            this.clientId = buildClientId(config.getString(ProducerConfig.CLIENT_ID_CONFIG), transactionalId);
+            String configuredClientId = config.getString(ProducerConfig.CLIENT_ID_CONFIG);
+
+            this.clientId = buildClientId(configuredClientId, transactionalId);
+
+            if (configuredClientId.isEmpty()) {
+                config.originals().put(ProducerConfig.CLIENT_ID_CONFIG, this.clientId);
```

Comment:
what's the usage for this? If user didn't supply a client.id, the printed configs will have client.id as empty field?

Reply:
yes, when users didn't config client.id, kafka clients will self generate it and the ProducerConfig values result will be: client.id =